Add support for `fill_value` and `set_empty_bucket_to` in BucketResampler `get_sum`

ameraner opened the pull request:
This PR adds support for `fill_value` and `set_empty_bucket_to` in BucketResampler `get_sum`. This is a nice feature, but most importantly it fixes the satpy `bucket_sum` interface.

Closes https://github.com/pytroll/satpy/issues/2805, closes https://github.com/pytroll/satpy/issues/1746, closes https://github.com/pytroll/pyresample/issues/354

 - [x] Tests added <!-- for all bug fixes or enhancements -->
 - [x] Tests passed <!-- for all non-documentation changes -->
 - [x] Passes ``git diff origin/main **/*py | flake8 --diff`` <!-- remove if you did not edit any Python files  -->
 - [x] Fully documented <!-- remove if this change should not be visible to users, e.g., if it is an internal clean-up, or if this is part of a larger project that will be documented later -->

djhoese commented on pyresample/bucket/__init__.py:
```diff
@@ -202,20 +202,30 @@ def _get_indices(self):
+            (similarly to Numpy's `sum`).
+            In both cases, empty buckets are set to `set_empty_bucket_to`.
+            Default: True
+        set_empty_bucket_to : float
```
Can we work on a better name for this? The "set" part is a verb and seems more like the name for a function than a variable. What about `empty_fill` or something else? Also in what cases would this parameter not be 0?

ameraner replied:
good point, thanks. `empty_fill` doesn't feel very self-explanatory to me though.. maybe `emtpy_bucket_value`?

In my use case, for example, I'm setting it to `np.nan` to maintain a sparse array after resampling, which is easier/lighter to handle and maintains transparency properties.

djhoese replied:
enpty_bucket_value is better but "value" seems redundant. Like calling a variable "var". But I'm not sure it is that big of a deal that it *needs* changing.

ameraner replied:
I was proposing `emtpy_bucket_value` since it goes along `fill_value`. But we can agree on `empty_bucket` :)

> But I'm not sure it is that big of a deal that it needs changing.

Do you mean the variable name or the entire concept of the new kwarg?

djhoese replied:
I meant that `empty_bucket_value` is probably fine and good point with `fill_value` (although that might be historic with CF `_FillValue` and similar). Anyway...`empty_bucket_value` sounds good.

ameraner replied:
Done, I also added a `if empty_bucket_value != 0:` check to avoid an unnecessary `where` call for the default case.

djhoese commented on pyresample/bucket/__init__.py:
```diff
@@ -228,8 +237,11 @@ def get_sum(self, data, skipna=True):
-        # Remove NaN values from the data when used as weights
-        weights = da.where(np.isnan(data), 0, data)
+        # Remove fill_values values from the data when used as weights
+        if np.isnan(fill_value):
+            weights = da.where(np.isnan(data), 0, data)
+        else:
+            weights = da.where(data == fill_value, 0, data)
```
Instead of the `if` statement, what about a helper for getting the invalid mask that does the nan check on the fill value and returns the appropriate mask? That way this stays `weights = da.where(invalid_mask, 0, data)`.

ameraner replied:
Done

djhoese commented on pyresample/bucket/__init__.py:
```diff
@@ -202,20 +202,29 @@ def _get_indices(self):
+            Note that a bucket could become 0 as the result of a sum
+            of positive and negative values. If the user needs to identify these zero-buckets reliably,
+            `get_count()` can be used for this purpose.
+            Default: np.nan
```
Why did you change the default to NaN instead of 0?

ameraner replied:
Sorry, that was a mistake in the docstring, well spotted. Fixed now.

djhoese commented on pyresample/bucket/__init__.py:
```diff
@@ -241,16 +253,22 @@ def get_sum(self, data, skipna=True):
-            nan_bins, _ = da.histogram(self.idxs[nans], bins=out_size,
-                                       range=(0, out_size))
-            statistic = da.where(nan_bins > 0, np.nan, statistic)
+            if np.isnan(fill_value):
+                missing_val = np.isnan(data)
+            else:
+                missing_val = data == fill_value
```
This could use that same mask helper function I mentioned above.

ameraner replied:
Done

djhoese commented on pyresample/test/test_bucket.py:
```diff
@@ -174,10 +174,61 @@ def test_get_sum_nan_data_skipna_true(self):
         self.assertEqual(np.count_nonzero(np.isnan(result)), 0)
         self.assertEqual(np.nanmin(result), 0)
 
+    def test_get_sum_non_default_fill_value_skipna_false(self):
+        """Test drop-in-a-bucket sum for data input with non-default fill_value and skipna=False."""
+        data = da.from_array(np.array([[2., 255], [5., 255]]),
+                             chunks=self.chunks)
+
+        result = self._get_sum_result(data, skipna=False, fill_value=255)
+        # 2 + fill_value  is fill_value, all-fill_value is fill_value
+        self.assertEqual(np.count_nonzero(result == 255), 2)
+        # 5 is untouched in a single bin
+        self.assertEqual(np.count_nonzero(result == 5.), 1)
+        # rest is 0
+        self.assertEqual(np.nanmin(result), 0)
```
These tests all seem very similar. Could you separate these tests into a pytest-based class and use parametrization to get all the cases covered in a single test? Bonus points if you update this existing class to pytest (resampler fixture, remove setup method, remove base class, use `assert` instead of `self.assertX`).

ameraner replied:
Done, and I like bonus points so I moved everything to pytest :)

ameraner commented:
I addressed all comments, and most importantly I switched all bucket tests to pytest. I realised there are many ways to do this (like do we need a base test class containing all tests?), so please let me know if you like it like this. I also used fixtures with module scope to optimise it a bit.

I also parametrised the get_sum arguments tests into one test as proposed. It took a while to figure out the combination and simplification of the test flow, but indeed I think it improved the test readability, and made the testing of more corner cases much easier. So thanks for pushing me to do it :)

djhoese commented on pyresample/bucket/__init__.py:
```diff
@@ -456,6 +469,14 @@ def get_fractions(self, data, categories=None, fill_value=np.nan):
         return results
 
 
+def get_invalid_mask(data, fill_value):
```
Small nitpick, maybe do a `_` prefix to this function to show that it isn't a public interface.

pnuu replied:
Yep, agreed.

ameraner replied:
Ok, done.

djhoese commented on pyresample/bucket/__init__.py:
```diff
@@ -202,20 +202,29 @@ def _get_indices(self):
         data : Numpy or Dask array
             Data to be binned and summed.
+        fill_value : float
+            Fill value to mark missing/invalid values in the input data.
```
The previous wording suggest to me on first reading that the `fill_value` was used to write/assign to some array, but if I'm understanding it correctly it is already existing in the input data so I thought maybe this could be less confusing:

```suggestion
            Fill value of the input data marking missing/invalid values.
```

ameraner replied:
indeed, fixed as proposed.

djhoese reviewed (changes requested):
Small suggestions, but otherwise looks good. I'd really like @pnuu to review this given he's the original author (right?).

pnuu reviewed (commented):
LGTM, just the one nitpick-y change that Dave pointed out.

djhoese commented:
@ameraner I'd like to make a release soon. Any idea if you'll have time to finish this PR up soon? Looks like it might just be small things left to change. Also, does this PR also close #364?

djhoese commented:
CC @zxdawn about comparisons to #364

ameraner commented:
Hi @djhoese , thanks for the review and the ping, indeed it would be great to have this in the next release. I have pushed the modifications as requested. 

I believe it should close the PR, yes, as it implements the prosposed functionalities, even with a better granularity using the `fill_value` and `empty_bucket_value` arguments.

djhoese commented on pyresample/test/test_bucket.py:
```diff
@@ -17,328 +17,377 @@
+    return da.from_array(np.array([[60., 60.00001], [60.2, 60.3]]), chunks=CHUNKS)
+
+
+@pytest.fixture(scope="module")
+def resampler(adef, lons, lats):
+    """Get initialised resampler for tests."""
+    return bucket.BucketResampler(adef, lons, lats)
```
Is it a problem to make this module scoped when it seems the resampler object caches stuff in the instance variables:

```python
        self.x_idxs = None
        self.y_idxs = None
        self.y_idxs = None
        self.idxs = None
        self.idxs = None
        self._get_indices()
        self._get_indices()
        self.counts = None
        self.counts = None
```

ameraner replied:
The cached variables are related to the indices computed from the input lat, lon and target areadef, which are also module-scoped fixtures. So in this setting I think this is fine. As a cross-check, I just switched to function-scoped and all tests still pass, so there is no cross-contamination. Also, the tests that need different inputs or patched/mocked components do not use the resampler fixture.

zxdawn commented:
I agree this can close my old PR because it considers both nan and fill_value. Thanks!
